example/calculate_pi/calculate_pi.cpp

```diff
@@ -77,28 +79,27 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   return package;
 }
 
-TaskStatus ComputeArea(MeshBlock *pmb) {
-  // compute 1/r0^2 \int d^2x in_or_out(x,y) over the block's domain
-  auto &rc = pmb->real_containers.Get();
-  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
-  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
-  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
-  auto &coords = pmb->coords;
-
-  ParArrayND<Real> &v = rc->Get("in_or_out").data;
-
-  Real area;
-  Kokkos::parallel_reduce(
-      "calculate_pi compute area",
-      Kokkos::MDRangePolicy<Kokkos::Rank<3>>(pmb->exec_space, {kb.s, jb.s, ib.s},
-                                             {kb.e + 1, jb.e + 1, ib.e + 1},
-                                             {1, 1, ib.e + 1 - ib.s}),
-      KOKKOS_LAMBDA(int k, int j, int i, Real &larea) {
-        larea += v(k, j, i) * coords.Area(parthenon::X3DIR, k, j, i);
-      },
-      area);
-  Kokkos::deep_copy(pmb->exec_space, v.Get(0, 0, 0, 0, 0, 0), area);
+TaskStatus ComputeAreas(std::vector<MeshBlock *> &blocks) {
+  IndexRange ib = blocks[0]->cellbounds.GetBoundsI(IndexDomain::interior);
+  IndexRange jb = blocks[0]->cellbounds.GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = blocks[0]->cellbounds.GetBoundsK(IndexDomain::interior);
 
+  for (auto pmb : blocks) {
```

Comment:
I notice that the infrastructure now uses `std::vector<MeshBlock*>` as an underlying data structure for the tasking. We may want to rethink this at some point. It should probably be consistent with @AndrewGaspar 's MR #193 . Of course, we may want to change `std::list` there to `std::vector` too.

Reply:
yeah that's not a bad idea.  actually it may be trivial to support both.

Reply:
For now I wouldn't worry about it. But we should keep it in mind. I think they may come into alignment at some point, because the performance bottleneck for `FindMeshBlock` I think benefits from `blocks` being a `std::vector` instead of a `std::list`.

Reply:
At least that's what I remember @jmstone saying?

Reply:
Yes, performance of FindMeshBlock is improved using a std::vector compared to std::list when there are very many MBs on a single rank.  Also beware of subtle bugs related to the way Athena++ uses pointers to parent objects (e.g. pmy_mblock) stored in classes when STL containers are used to store the MB list instead of the original hand-rolled linked list.  It seems the address of a pointer to an object added to an STL container is not reliable until the container is fully constructed.  This caused a bug it took me several days to find in my new test code.  Perhaps the container class used in Parthenon avoids this.

Reply:
> For now I wouldn't worry about it. But we should keep it in mind. I think they may come into alignment at some point, because the performance bottleneck for `FindMeshBlock` I think benefits from `blocks` being a `std::vector` instead of a `std::list`.

I've run into the same problem before, some of the stl containers are safer than others, particulalry the node based containers, pointers and iterators to items in a list, map, set, multimap and multiset are guaranteed to remain valid as long as the element is not removed from the container. Like @jmstone alluded too, if you add an element to a vector and it has to grow the memory then all the items may get copied to a new location, the same can happen for unordered maps and the other stl containers.

Reply:
Yes, it makes sense. so `std::vector<T>::push_back` might copy elements. 

I guess this means replacing elements in a vector with `std::shared_ptr` would also fix the problem, as the pointers are what would get copied.

Reply:
`std::list` is fine in terms of iterator invalidation. By design, adding or removing elements to a `std::list` doesn't invalidate iterators (or pointers to stored objects) unless you remove the object the iterator is pointing to.

To get a `std::vector` without pointer invalidation, just do `std::vector<std::unique_ptr<T>>`